README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# cloneRate <a href="https://bdj34.github.io/cloneRate/"><img src="man/figures/logo.png" align="right" height="136" /></a>

<!-- badges: start -->

[![R-CMD-check](https://github.com/bdj34/cloneRate/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/bdj34/cloneRate/actions/workflows/R-CMD-check.yaml)
[![Codecov test
coverage](https://codecov.io/gh/bdj34/cloneRate/branch/main/graph/badge.svg)](https://app.codecov.io/gh/bdj34/cloneRate?branch=main)
<!-- badges: end -->

The goal of cloneRate is to provide easily accessible methods for
estimating the growth rate of clones. The input should either be an
ultrametric phylogenetic tree with edge lengths corresponding to time,
or a non-ultrametric phylogenetic tree with edge lengths corresponding
to mutation counts. This package provides the internal lengths and
maximum likelihood methods for ultrametric trees and the shared
mutations method for mutation-based trees, all of which are from our
recent paper [cloneRate: fast estimation of single-cell clonal dynamics
using coalescent
theory](https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btad561/7271182).
We also provide a birth-death Markov Chain Monte Carlo (MCMC) approach
using the probability density derived in Eq. 5 of [Tanja Stadler’s
work](https://pubmed.ncbi.nlm.nih.gov/19631666/).

To test our methods, we provide a fast way to simulate the coalescent
(tree) of a sample from a birth-death branching process, which is a
direct result of [Amaury Lambert’s
work](https://pubmed.ncbi.nlm.nih.gov/29704514/).

## Installation

Install from CRAN with the following:

``` r
install.packages("cloneRate")
```

Alternatively, you can install the development version of cloneRate from
[GitHub](https://github.com/). For this basic tutorial and our
vignettes, we will also use a few other packages, which can all be
installed from CRAN. Because these are listed as packages we ‘suggest’,
running the following command (with `dependencies = TRUE`) will install
them along with the vignettes:

``` r
# Install devtools if you don't have it already
install.packages(setdiff("devtools", rownames(installed.packages())))

# Install
devtools::install_github("bdj34/cloneRate", build_vignettes = TRUE, dependencies = TRUE)
```

Alternatively, you can install them manually:

``` r
install.packages(setdiff(c("ggplot2", "ggsurvfit", "survival", "car"), rownames(installed.packages())))
```

## Example

We’ll walk through simulating a single tree and plotting it. Then we’ll
apply our growth rate methods.

### Simulate data

We can simulate a sample of size n from a birth-death tree as follows:

``` r
library(cloneRate)
library(ape)
library(ggplot2)

# Generate a sampled tree with 100 tips from a 20 year birth-death process with birth rate a=1 and death rate b=0.5
tree <- simUltra(a = 1, b = 0.5, cloneAge = 40, n = 100)
```

Now that we have simulated the tree, let’s plot it:

``` r
# Plot, then add scale and title
plot.phylo(tree,
  direction = "downwards",
  show.tip.label = FALSE, edge.width = 2
)
axisPhylo(side = 2, backward = FALSE, las = 1)
title(main = "Simulated ultrametric tree", ylab = "Time (years)")
```

<img src="man/figures/README-plotTree-1.png" width="100%" />

### Estimate growth rate of one tree

We can use this tree as input to our methods for growth rate estimation:

``` r
# Estimate the growth rate r=a-b=0.5 using maximum likelihood
maxLike.df <- maxLikelihood(tree)
print(paste0("Max. likelihood estimate = ", round(maxLike.df$estimate, 3)))
#> [1] "Max. likelihood estimate = 0.537"

# Estimate the growth rate r=a-b=0.5 using internal lengths
intLengths.df <- internalLengths(tree)
print(paste0("Internal lengths estimate = ", round(intLengths.df$estimate, 3)))
#> [1] "Internal lengths estimate = 0.527"
```

Because we’re simulating a new tree each time, the estimate will change
with each run, so don’t be worried if your results don’t match exactly.

### Estimate growth rate of many trees

In [our
paper](https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btad561/7271182),
we use simulated trees to test our growth rate estimates. As an example,
let’s load some simulated data that comes with our package,
`exampleUltraTrees` has 100 ultrametric trees. In the “metadata”
data.frame we will find the ground truth growth rate, which in this case
is 1. Let’s apply our methods to all 100 trees.

``` r
# Here we are applying two methods to all of the ultrametric trees
resultsUltraMaxLike <- maxLikelihood(exampleUltraTrees)
resultsUltraLengths <- internalLengths(exampleUltraTrees)
```

Notice how the functions `maxLikelihood()` and `internalLengths()` can
take as input either a single tree or a list of trees. Either way, the
output is a `data.frame` containing the results. Now that we have 100
estimates on 100 different trees from 2 different methods, let’s plot
the distributions

``` r
# Combine all into one df for plotting. This works because the columns are the same
resultsCombined <- rbind(resultsUltraMaxLike, resultsUltraLengths)

# Plot, adding a vertical line at r=1 because that's the true growth rate
ggplot(resultsCombined) +
  geom_density(aes(x = estimate, color = method), linewidth = 1.5) +
  geom_vline(xintercept = exampleUltraTrees[[1]]$metadata$r) +
  theme_bw() +
  theme(
    axis.text.y = element_blank(), axis.ticks.y = element_blank(),
    legend.title = element_blank()
  ) +
  xlab("Net growth rate estimate (r)") +
  ylab("Density") +
  scale_color_manual(labels = c("Internal lengths", "Max. likelihood"), values = c("black", "#009E73"))
```

<img src="man/figures/README-plotExample-1.png" width="100%" />

Finally, let’s compute the root mean square error (RMSE) of the
estimates. We expect maximum likelihood to perform the best by RMSE, but
100 is a relatively small sample size so anything could happen…

``` r
# Calculate the RMSE
groundTruth <- exampleUltraTrees[[1]]$metadata$r[1]
rmse <- unlist(lapply(
  split(resultsCombined, resultsCombined$method),
  function(x) {
    sqrt(sum((x$estimate - groundTruth)^2) / length(x$estimate))
  }
))

print(rmse)
#>    lengths    maxLike 
#> 0.09119116 0.08756432
```

As expected, maximum likelihood performs the best. Note that this may
change if we regenerate the data. For more details, see the [cloneRate
website](https://bdj34.github.io/cloneRate/) or vignettes:

``` r
vignette("cloneRate-dataAnalysis", package = "cloneRate")
vignette("cloneRate-simulate", package = "cloneRate")
```

## References

Simulating the birth-death trees is a direct result of the work of
Amaury Lambert in:

- [Lambert, 2018](https://pubmed.ncbi.nlm.nih.gov/29704514/)

Our package comes with 42 clones annotated from four distinct
publications, which are the ones we use in our analysis. Note that there
are three clones profiled at two different timepoints, meaning there are
39 unique clones. The papers which generate this data are:

- [Williams et al. 2022](https://pubmed.ncbi.nlm.nih.gov/35058638/)
- [Mitchell et al. 2022](https://pubmed.ncbi.nlm.nih.gov/35650442/)
- [Fabre et al. 2022](https://pubmed.ncbi.nlm.nih.gov/35650444/)
- [Van Egeren et al. 2021](https://pubmed.ncbi.nlm.nih.gov/33621486/)

The birth-death MCMC (not shown here) is based on the probability
density derived in Eq. 5 of Tanja Stadler’s work:

- [Stadler, 2009](https://pubmed.ncbi.nlm.nih.gov/19631666/)

The mathematical basis for our estimates is detailed in full in [our
paper](https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btad561/7271182/).


NEWS.md
# cloneRate (development version)

# cloneRate 0.2.3

* Minor changes for CRAN resubmission

# cloneRate 0.2.2

* Wrap long-running examples in \donttest{} for CRAN

# cloneRate 0.2.1

* Change to CRAN installation instructions in README

# cloneRate 0.2.0

## Update for CRAN compatibility

* Make birth-death MCMC available to all users of package
* Move rstan from suggests to imports
* Update rstan methods to be consistent with rstantools recommendations

# cloneRate 0.1.0

## Add birth-death MCMC

* Add birth death MCMC targeting likelihood in equation 5 of Stadler 2009 "On incomplete sampling under birth–death models and connections to the sampling-based coalescent"
* Implements MCMC using rstan package, which is now listed in suggests

# cloneRate 0.0.1

* Added a `NEWS.md` file to track changes to the package.


R/data.R
#' Real clone data from human blood
#'
#' 42 clones (39 distinct) from 32 individual donors, 13 of whom have a diagnosis
#' of Myeloproliferative Neoplasm
#'
#' @docType data
#'
#' @usage data(realCloneData)
#'
#' @format A \code{list} of containing one \code{list} with the full ultrametric
#' trees from 30 of the 32 individual donors (the two from Van Egeren are not included),
#' and one \code{list} containing the 42 clone trees.In three cases, there are two
#' timepoints from the same clone, and these are separate phylo objects.
#' Each \code{list} contains a tree as a class \code{phylo} object.
#' See ape package documentation for details on class \code{phylo} objects.
#' Names of each \code{phylo} object (tree) in the list matches the naming used
#' in the sources and also includes driver, age, and clone number.
#'
#'
#' @references These datasets were generated and annotated in:
#' [Williams et al. 2022](https://pubmed.ncbi.nlm.nih.gov/35058638/)
#' [Mitchell et al. 2022](https://pubmed.ncbi.nlm.nih.gov/35650442/)
#' [Fabre et al. 2022](https://pubmed.ncbi.nlm.nih.gov/35650444/)
#' [Van Egeren et al. 2021](https://pubmed.ncbi.nlm.nih.gov/33621486/)
#' @keywords phylogenetics, hematopoiesis
#' @examples
#' # Plot full reconstructed tree from donor PD34493
#' ape::plot.phylo(cloneRate::realCloneData[["fullTrees"]][["PD34493"]],
#'   direction = "downwards", show.tip.label = FALSE
#' )
#'
#' @source <https://pubmed.ncbi.nlm.nih.gov/35058638/>
#' @source <https://pubmed.ncbi.nlm.nih.gov/35650442/>
#' @source <https://pubmed.ncbi.nlm.nih.gov/35650444/>
#' @source <https://pubmed.ncbi.nlm.nih.gov/33621486/>
"realCloneData"







#' Example ultrametric tree data
#'
#' Set of 100 time-based ultrametric trees reconstructed from the distribution
#' of a sample of n=100 tips.
#' All trees have a net growth rate of 1 with birth rates between 1 and 2
#' (sampled from a uniform distribution). Death rates are equal to the chosen
#' birth rate minus 1. Tree reconstruction uses the exact distribution of
#' coalescence times described in "The coalescent of a sample from a binary
#' branching process", Lambert A., Theor. Pop. Bio. 2018. Tree construction and
#' formatting uses \code{ape} R package [ape::rcoal()].
#'
#' @docType data
#'
#' @usage data(exampleUltraTrees)
#'
#' @format A \code{list} of objects of class \code{phylo}
#' \describe{
#'  \item{edge}{A matrix of edge connections which reconstruct the tree.}
#'  \item{edge.length}{A numeric vector of the branch lengths of the connections
#'  in \code{edge} matrix. Units are years.}
#'  \item{tip.label}{A character vector containing the  (arbitrary in this case)
#'   labels for the 100 tips/samples of the tree.}
#'  \item{Nnode}{Integer number of internal nodes of the tree}
#'  \item{params}{\code{data.frame} containing info on the params used to generate
#'  the tree}
#'  See ape package for details on class \code{phylo} objects.
#' }
#' @references This data set was created for the cloneRate package using
#' coalescent theory approaches described in "The coalescent of a sample from a
#' binary branching process", Lambert A., Theor. Pop. Bio. 2018.
#' @keywords phylogenetics, birth-death trees, Coalescent Point Process.
#' @examples
#' # Plot first of 100 trees
#' ape::plot.phylo(cloneRate::exampleUltraTrees[[1]],
#'   direction = "downwards", show.tip.label = FALSE
#' )
"exampleUltraTrees"







#' Example mutation tree data
#'
#' Set of 100 mutation based trees reconstructed from the distribution
#' of a sample of n=100 tips.
#' All trees have a net growth rate of 1 with birth rates between 1 and 2
#' (sampled from a uniform distribution). Death rates are equal to the chosen
#' birth rate minus 1. Tree reconstruction uses the exact distribution of
#' coalescence times described in "The coalescent of a sample from a binary
#' branching process", Lambert A., Theor. Pop. Bio. 2018. Tree construction and
#' formatting uses \code{ape} R package [ape::rcoal()]. We then change the edge
#' lengths from time-based to mutation-based by drawing from a poisson
#' distribution with mean equal to edge length (in units of time) multiplied
#' by the mutation rate, nu, which is drawn from a uniform distribution between
#' 10 and 20 mutations per year.
#'
#' @docType data
#'
#' @usage data(exampleMutTrees)
#'
#' @format A \code{list} of objects of class \code{phylo}
#' \describe{
#'  \item{edge}{A matrix of edge connections which reconstruct the tree.}
#'  \item{edge.length}{A numeric vector of the branch lengths of the connections
#'  in \code{edge} matrix. Units are mutations.}
#'  \item{tip.label}{A character vector containing the  (arbitrary in this case)
#'   labels for the 100 tips/samples of the tree.}
#'  \item{Nnode}{Integer number of internal nodes of the tree}
#'  \item{params}{\code{data.frame} containing info on the params used to generate
#'   the tree}
#'  See ape package for details on class \code{phylo} objects.
#' }
#' @references This data set was created for the cloneRate package using
#' coalescent theory approaches described in "The coalescent of a sample from a
#' binary branching process", Lambert A., Theor. Pop. Bio. 2018.
#' @keywords phylogenetics, birth-death trees, Coalescent Point Process
#' @examples
#' # Plot first of 100 trees
#' ape::plot.phylo(cloneRate::exampleMutTrees[[1]],
#'   direction = "downwards", show.tip.label = FALSE
#' )
#'
"exampleMutTrees"





#' Longitudinal validation data
#'
#' For three individuals with clonal expansions that can be estimated using our methods,
#' we have longitudinal data to orthogonally validate these estimates, which is included here.
#' Additionally, for 13 clones with a driver gene matching a driver gene in the single cell
#' data, but without a match to a specific clone, we include this longitudinal data as well.
#'
#' @format A \code{data.frame} containing all the information needed
#' \describe{
#'  \item{Sample.ID}{The individual's ID}
#'  \item{Age}{Individual's age at the various sampling times}
#'  \item{VAF}{The variant allele frequency at the various sampling times for the clone of interest}
#'  \item{Gene}{Gene or genes with mutation that identifies the clone}
#'  \item{Protein}{Protein affected by the mutation}
#'  \item{cellType}{The type of cells used for sequencing}
#'  \item{cloneName}{The name we use for the clone to match to single cell data, if applicable.}
#' }
#'
#' @references These datasets were generated and annotated in:
#' [Williams et al. 2022](https://pubmed.ncbi.nlm.nih.gov/35058638/)
#' [Fabre et al. 2022](https://pubmed.ncbi.nlm.nih.gov/35650444/)
#'
#' @keywords longitudinal
#'
#' @examples
#' library(ggplot2)
#' # Plot longitudinal data from PD9478
#' ggplot(longitudinalData[longitudinalData$Sample.ID == "PD9478", ]) +
#'   geom_point(aes(x = Age, y = VAF))
"longitudinalData"
